engine/servermaster/jobop/backoff.go

```diff
@@ -0,0 +1,191 @@
+// limitations under the License.
+
+package jobop
+
+import (
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/engine/pkg/clock"
+	"go.uber.org/zap"
+)
+
+type backoffEventType int32
+
+const (
+	backoffOnline backoffEventType = iota + 1
+	backoffOffline
+)
+
+const (
+	defaultBackoffInitInterval = 10 * time.Second
+	defaultBackoffMaxInterval  = 5 * time.Minute
+	defaultBackoffMultiplier   = 2.0
+	// If a job can keep running for more than 10 minutes, it won't be backoff
+	// If a job keeps failing, the max back interval is 5 minutes, and 10 minutes
+	// can keep at least one failed record.
+	defaultBackoffRestInterval = 2 * defaultBackoffMaxInterval
+)
+
+type backoffOpts struct {
+	clocker         clock.Clock
+	resetInterval   time.Duration
+	initialInterval time.Duration
+	maxInterval     time.Duration
+	multiplier      float64
+}
+
+// BackoffOption is used to set options to job backoff
+type BackoffOption func(*backoffOpts)
+
+// WithResetInterval sets resetInterval to backoffOpts
+func WithResetInterval(interval time.Duration) BackoffOption {
+	return func(opts *backoffOpts) {
+		opts.resetInterval = interval
+	}
+}
+
+// WithInitialInterval sets initialInterval to backoffOpts
+func WithInitialInterval(interval time.Duration) BackoffOption {
+	return func(opts *backoffOpts) {
+		opts.initialInterval = interval
+	}
+}
+
+// WithMaxInterval sets maxInterval to backoffOpts
+func WithMaxInterval(interval time.Duration) BackoffOption {
+	return func(opts *backoffOpts) {
+		opts.maxInterval = interval
+	}
+}
+
+// WithMultiplier sets multiplier to backoffOpts
+func WithMultiplier(multiplier float64) BackoffOption {
+	return func(opts *backoffOpts) {
+		opts.multiplier = multiplier
+	}
+}
+
+// WithClocker sets clocker to backoffOpts
+func WithClocker(clocker clock.Clock) BackoffOption {
+	return func(opts *backoffOpts) {
+		opts.clocker = clocker
+	}
+}
+
+type backoffEvent struct {
+	tp backoffEventType
+	ts time.Time
+}
+
+// NewJobBackoff creates a new job backoff
+func NewJobBackoff(jobID string, options ...BackoffOption) *JobBackoff {
+	opts := &backoffOpts{
+		resetInterval:   defaultBackoffRestInterval,
+		initialInterval: defaultBackoffInitInterval,
+		maxInterval:     defaultBackoffMaxInterval,
+		multiplier:      defaultBackoffMultiplier,
+		clocker:         clock.New(),
+	}
+	for _, option := range options {
+		option(opts)
+	}
+	b := &JobBackoff{
+		jobID: jobID,
+		opts:  opts,
+	}
+	errBackoff := backoff.NewExponentialBackOff()
+	errBackoff.InitialInterval = opts.initialInterval
+	errBackoff.MaxInterval = opts.maxInterval
+	errBackoff.Multiplier = opts.multiplier
+	errBackoff.Reset()
+	b.errBackoff = errBackoff
+
+	return b
+}
+
+// JobBackoff is a job backoff manager, it recoreds job online and offline events
+// and determines whether a job can be re-created based on backoff mechanism.
+type JobBackoff struct {
+	jobID           string
+	events          []backoffEvent
```

Comment:
I don't really get the point of this `event queue`.
In my mind, we always get the seqence `OnWorkerOnline, OnWorkerOffline, OnWorkerOnline, OnWorkerOffline...`, so every `OnWorkerOffline` will update the `backoffInterval`, a long `online` interval will reset the `backoffInterval`.

So a `lastOnlineTs`  seems enough for me.

Reply:
The sequence could be `Offline`, `Offline`, `Offline` without any Online.
Currently implement can only record the latest online ts, but it is flexible to record all events.